src/ape_cache/query.py

```diff
@@ -0,0 +1,337 @@
+
+    @estimate_query_clause.register
+    def transaction_estimate_query_clause(self, query: BlockTransactionQuery) -> TextClause:
+        return text(
+            """
+    SELECT COUNT(*)
+    FROM transactions
+    WHERE transactions.block_hash = :block_hash
+        """
+        ).bindparams(block_hash=query.block_id)
+
+    @estimate_query_clause.register
+    def contract_events_estimate_query_clause(self, query: ContractEventQuery) -> TextClause:
+        return text(
+            """
+    SELECT COUNT(*)
+    FROM contract_events
+    WHERE contract_events.block_number >= :start_block
+    AND contract_events.block_number <= :stop_block
+    AND contract_events.block_number % :step = 0
+        """
+        ).bindparams(start_block=query.start_block, stop_block=query.stop_block, step=query.step)
+
+    @singledispatchmethod
+    def compute_estimate(self, result: CursorResult, query: QueryType) -> Optional[int]:
+        return None  # can't handle this query
+
+    @compute_estimate.register
+    def compute_estimate_block_query(
+        self,
+        query: BlockQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.rowcount == (query.stop_block - query.start_block) // query.step:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        # TODO: Allow partial queries
+        return None
+
+    @compute_estimate.register
+    def compute_estimate_block_transaction_query(
+        self,
+        query: BlockTransactionQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.rowcount > 0:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        return None
+
+    @compute_estimate.register
+    def compute_estimate_contract_events_query(
+        self,
+        query: ContractEventQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.rowcount == (query.stop_block - query.start_block) // query.step:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        # TODO: Allow partial queries
+        return None
+
+    def estimate_query(self, query: QueryType) -> Optional[int]:
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self.estimate_query_clause(query))
+
+                if not result:
+                    return None
+
+                return self.compute_estimate(result, query)
+        except QueryEngineError as err:
+            logger.warning(f"Cannot perform query on cache database: {err}")
+            return None
+
+    # Fetch data
+    @singledispatchmethod
+    def perform_query_clause(self, query: QueryType) -> TextClause:
+        raise QueryEngineError(
+            """
```

Comment:
When you put the quotes like this, it will include the leading and trailing `\n` characters, so the exception is going to look like:

```
ERROR:
Not a compatible QueryType. For more details see our docs
https://docs.apeworx.io/ape/stable/methoddocs/exceptions.html#ape.exceptions.QueryEngineError
```

I am guessing the intention here is to do something like this instead:

```python
raise QueryEngineError(
    "Not a compatible QueryType. For more details see our docs "
    "https://docs.apeworx.io/ape/stable/methoddocs/exceptions.html#ape.exceptions.QueryEngineError"
)
```

which will cause:

```
ERROR: Not a compatible QueryType. For more details see our docs https://docs.apeworx.io/ape/stable/methoddocs/exceptions.html#ape.exceptions.QueryEngineError
```

Reply:
[ecb2649](https://github.com/ApeWorX/ape/pull/680/commits/ecb264926c0999b33af86dba3d4643e23b1b2fe5)